File: cycle_detection/py/cycle_detection.py
# Floyd's tortoise and hare algorithm
# Time:  O(n)
# Space: O(1)

def cycle_detection(head):
    fast = slow = head
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        if fast == slow:
            return True
    return False

def cycle_finding(head):
    fast = slow = head
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        if fast == slow:
            slow = head
            while fast != slow:
                slow = slow.next
                fast = fast.next
            return fast
    return None

File: linked_list/cxx/cycle_detection.cxx
#include "LinkedList.h"

bool cycle_detection(ListNode* head)
{
	ListNode *slow = head, *fast = head;
	while(fast and fast->next)
	{
		slow = slow->next;
		fast = fast->next->next;
		if(slow == fast)
			return true;
	}
	return false;
}

ListNode* cycle_finding(ListNode* head)
{
	ListNode *slow = head, *fast = head;
	while(fast and fast->next)
	{
		slow = slow->next;
		fast = fast->next->next;
		if(slow == fast)
		{
			slow = head;
			while(slow != fast)
			{
				slow = slow->next;
				fast = fast->next;
			}
			return fast;
		}
	}
	return NULL;
}
